File: sdk/samples/all in on code/Visual Studio 2008/CSWebBrowserWithProxy/InternetProxy.cs
/****************************** Module Header ******************************\
 Module Name:  InternetProxy.cs
 Project:      CSWebBrowserWithProxy
 Copyright (c) Microsoft Corporation.
 
 This class is used to describe a proxy server and the credential to connect to it.
 Please set the proxy servers in ProxyList.xml first.
 
 This source is subject to the Microsoft Public License.
 See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
 All other rights reserved.
 
 THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
 EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
 WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\***************************************************************************/

using System.Collections.Generic;
using System.Xml.Linq;

namespace CSWebBrowserWithProxy
{
    public class InternetProxy
    {
        public string ProxyName { get; set; }
        public string Address { get; set; }
        public string UserName { get; set; }
        public string Password { get; set; }

        public override bool Equals(object obj)
        {
            if (obj is InternetProxy)
            {
                InternetProxy proxy = obj as InternetProxy;

                return this.Address.Equals(proxy.Address, System.StringComparison.OrdinalIgnoreCase) 
                    && this.UserName.EndsWith(proxy.UserName, System.StringComparison.OrdinalIgnoreCase)
                    && this.Password.Equals(proxy.Password, System.StringComparison.Ordinal);
            }
            else
            {
                return false;
            }          
        }

        public override int GetHashCode()
        {
            return this.Address.GetHashCode()
                +this.UserName.GetHashCode()
                +this.Password.GetHashCode();
        }

        public static readonly InternetProxy NoProxy = new InternetProxy
        {
            Address = string.Empty,
            Password = string.Empty,
            ProxyName = string.Empty,
            UserName = string.Empty
        };
        
        static List<InternetProxy> proxyList = null;
        public static List<InternetProxy> ProxyList
        {
            get
            {
                // Gets the proxy servers in ProxyList.xml.
                if (proxyList == null)
                {
                    proxyList = new List<InternetProxy>();
                    try
                    {
                        XElement listXml = XElement.Load("ProxyList.xml");
                        foreach (var proxy in listXml.Elements("Proxy"))
                        {
                            proxyList.Add(
                                new InternetProxy
                                {
                                    ProxyName = proxy.Element("ProxyName").Value,
                                    Address = proxy.Element("Address").Value,
                                    UserName = proxy.Element("UserName").Value,
                                    Password = proxy.Element("Password").Value
                                });
                        }
                    }
                    catch (System.Exception)
                    {
                    }                  
                }
                return proxyList;
            }
        }
    }
}

File: sdk/samples/all in on code/Visual Studio 2008/VBWebBrowserWithProxy/InternetProxy.vb
' *************************** Module Header ******************************\
' Module Name:  InternetProxy.vb
' Project:      VBWebBrowserWithProxy
' Copyright (c) Microsoft Corporation.
' 
' This class is used to describe a proxy server and the credential to connect to it.
' Please set the proxy servers in ProxyList.xml first.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
' **************************************************************************

Public Class InternetProxy

    Dim _proxyName As String
    Public Property ProxyName() As String
        Get
            Return _proxyName
        End Get
        Set(ByVal value As String)
            _proxyName = value
        End Set
    End Property

    Dim _address As String
    Public Property Address() As String
        Get
            Return _address
        End Get
        Set(ByVal value As String)
            _address = value
        End Set
    End Property

    Dim _userName As String
    Public Property UserName() As String
        Get
            Return _userName
        End Get
        Set(ByVal value As String)
            _userName = value
        End Set
    End Property

    Dim _password As String
    Public Property Password() As String
        Get
            Return _password
        End Get
        Set(ByVal value As String)
            _password = value
        End Set
    End Property

    Public Overloads Overrides Function Equals(ByVal obj As Object) As Boolean
        If TypeOf obj Is InternetProxy Then
            Dim proxy As InternetProxy = TryCast(obj, InternetProxy)

            Return Me.Address.Equals(proxy.Address, System.StringComparison.OrdinalIgnoreCase) _
                AndAlso Me.UserName.EndsWith(proxy.UserName, System.StringComparison.OrdinalIgnoreCase) _
                AndAlso Me.Password.Equals(proxy.Password, System.StringComparison.Ordinal)
        Else
            Return False
        End If
    End Function

    Public Overrides Function GetHashCode() As Integer
        Return Me.Address.GetHashCode() _
            + Me.UserName.GetHashCode() _
            + Me.Password.GetHashCode()
    End Function

    Public Shared ReadOnly NoProxy As InternetProxy = New InternetProxy With {.Address = String.Empty, .Password = String.Empty, .ProxyName = String.Empty, .UserName = String.Empty}

    Private Shared _proxyList As List(Of InternetProxy) = Nothing
    Public Shared ReadOnly Property ProxyList() As List(Of InternetProxy)
        Get
            ' Gets the proxy servers in ProxyList.xml.
            If _proxyList Is Nothing Then
                _proxyList = New List(Of InternetProxy)()
                Try
                    Dim listXml As XElement = XElement.Load("ProxyList.xml")
                    For Each proxy In listXml.Elements("Proxy")
                        _proxyList.Add(New InternetProxy With _
                                       {.ProxyName = proxy.Element("ProxyName").Value, _
                                        .Address = proxy.Element("Address").Value, _
                                        .UserName = proxy.Element("UserName").Value, _
                                        .Password = proxy.Element("Password").Value})
                    Next proxy
                Catch e1 As System.Exception
                End Try
            End If
            Return _proxyList
        End Get
    End Property
End Class